libraries/DataFlash/DataFlash_File.cpp

```diff
@@ -827,7 +827,7 @@ uint16_t DataFlash_File::get_num_logs()
 void DataFlash_File::stop_logging(void)
 {
     // best-case effort to avoid annoying the IO thread
-    const bool have_sem = write_fd_semaphore->take(1);
+    const bool have_sem = write_fd_semaphore->take(10000);
```

Comment:
1 ms is not enough. Write operation in _io_timer function can take up to several hundreds on ms.

Reply:
This is on the main thread.  It seems like a bad idea to stop flying the vehicle for 10 seconds in the event the IO thread is holding this semaphore :-)

Reply:
Thank you for review!
This is not as bad, as it seems at first glance. Start/stop logging is not called in armed state: only just after arming and before disarming. Other cases excluded, cause log mavlink commands are rejected in armed state. Also this code could hit similar timeout in file operations ( ::close ) anyway.

Reply:
Couple of things.  Firstly, we do call this function when we run out of space on the SD card.

Secondly - we can't ever stop the main thread for a prolonged period of time or the EKF gets very shirty.  Also, since we'll stop sending heartbeats to the GCS the user will believe their vehicle is crashed.

You've highlighted a real problem here, of course.  There's a couple of ways it could be fixed.  

One would be to move all file operations (open/close/stat etc) into the IO thread.  That's a lot of work as a lot of data will have to be passed back and forth between the threads.

The second is to make `stop_logging` on the backends `bool` and handle the case more elegantly where required - with a retry or erroring out.

Reply:
When we run out of space on the SD card, stop_logging is called from _io_timer function, so main thread is not affected. I agree that 10s is a extream value, but 1s should enouth to fix issue and safe to use. 

I was thinking about both of you suggestions, but I can not find a way to inplement them. Moving all filesystem io to separate thread would be great, but it require a lot of rework of all Dataflash lib. Returning fail error flag from stop_logging is not great, cause it is not error really - so some things will fail without any reason.